### PR TITLE
Change implementation of softmax cross entropy loss 

### PR DESCRIPTION
**Pull request by xiejw**
This PR changed two things. 

1. Used a numerical stable implementation for softmax cross entropy.
2. Changed the gradient wr.t. Now the loss only generates the gradient w.r.t. logits. This is consistent with its variant (Int label version).

In addition, added numerical tests to verify the correctness.

### CODE REVIEW

File: Sources/DeepLearning/Loss.swift [hunk 1]
```diff
@@ -53,11 +53,20 @@ func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
...
+    return Raw.softmaxCrossEntropyWithLogits(features: logits, labels: oneHotLabels).loss.mean()
+}
+
+@usableFromInline
+func _vjpSoftmaxCrossEntropy<Scalar: TensorFlowFloatingPoint>(
+    logits: Tensor<Scalar>, oneHotLabels: Tensor<Scalar>
+) -> (Tensor<Scalar>, (Tensor<Scalar>) -> Tensor<Scalar>) {
+    let (loss, grad) = Raw.softmaxCrossEntropyWithLogits(features: logits, labels: oneHotLabels)
+    let batchSize = Tensor<Scalar>(logits.shapeTensor[0])
+    return (loss.mean(), { v in (v / batchSize) * grad })
```
Comment (rxwei) [inline Sources/DeepLearning/Loss.swift]: Unnecessary parens?
```suggestion
    return (loss.mean(), { v in v / batchSize * grad })
```

File: Tests/DeepLearningTests/LossTests.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            at: logits,
+            in: { softmaxCrossEntropy(logits: $0, oneHotLabels: labels) })
+        assertAllClose(expected: expectedGradients, got: gradients)
+    }
+
+    func assertAllClose(expected a: Tensor<Float>, got b: Tensor<Float>, tol: Float = 1e-6) {
+        XCTAssertEqual(a.shape, b.shape)
+        for (index, elementInA) in a.scalars.enumerated() {
+            let elementInB = b.scalars[index]
+            XCTAssertTrue(
```
Comment (rxwei) [inline Tests/DeepLearningTests/LossTests.swift]: Have you considered using `XCTAssertLessThan`?

File: Tests/DeepLearningTests/LossTests.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        // As the loss is mean loss, we should scale the golden loss numbers.
+        let expectedGradients = expectedGradientsBeforeMean / Float(logits.shape[0])
+        let gradients = gradient(
+            at: logits,
+            in: { softmaxCrossEntropy(logits: $0, oneHotLabels: labels) })
+        assertAllClose(expected: expectedGradients, got: gradients)
+    }
+
+    func assertAllClose(expected a: Tensor<Float>, got b: Tensor<Float>, tol: Float = 1e-6) {
```
Comment (rxwei) [inline Tests/DeepLearningTests/LossTests.swift]: Sorry for the nitpick on the test API.
* Consider spelling out argument labels for clarity.
* How about calling it `assertElementsEqual`? The `tolerance:` parameter already indicates it's an approximation.
* It's probably longer than 100 columns after my suggested edit.

```suggestion
    func assertAllClose(expected: Tensor<Float>, actual: Tensor<Float>, tolerance: Float = 1e-6) {
```

File: Tests/DeepLearningTests/LossTests.swift (hunk 2 above)
Comment (xiejw) [reply Tests/DeepLearningTests/LossTests.swift]: Done. Sounds better.
